feature-cmake-object-libraries: cellular+netsocket as OBJECT

0xc0170 opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

I wanted to touch only cellular library but due to dependencies in the tree, I had to touch many more. This practically refactors cellular, mbedtls, nanostack and netsockets plus additional libraries in the tree that are required.

2 libraries  that are taken from upstream are built as STATIC due to their nature - they are built even in the upstream as static. The rest are OBJECT libraries and INTERFACE for sublibraries like drivers.

I am able to build cellular example, I'll start testing netsockets now.

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->


#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [x] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

0xc0170 committed:
CMake cellular: set mbed-cellular as OBJECT library

0xc0170 committed:
CMake cellular: create INTERFACE library stmod

0xc0170 committed:
CMake cellular: refactor drivers to be CMake targets

0xc0170 committed:
CMake cellular: fix n2xx library name

0xc0170 committed:
CMake cellular: add rtos as dependency

0xc0170 committed:
CMake cellular: link modem privately to be part only of the object lib

0xc0170 committed:
CMake cellular: fix where we link, only one place it can be

0xc0170 committed:
CMake cellular: drivers are included from root

0xc0170 committed:
CMake mbedtls: remove unused CMakelist

0xc0170 committed:
CMake emac: refactor drivers to be own CMake targets

0xc0170 committed:
CMake mbedtls: refactor drivers to be own CMake targets

0xc0170 committed:
CMake: connectivity libraries set to OBJECTS

0xc0170 committed:
CMake connectivity: all folders are optional

0xc0170 committed:
CMake cellular: fix drivers linking errors

0xc0170 committed:
CMake randlib: this can be static library

0xc0170 committed:
CMake libraries: do not always include them

0xc0170 committed:
CMake coap: this library can be static

0xc0170 committed:
CMake nanostack libservice: fix core depdency

0xc0170 committed:
CMake lwipstack: remove comment and linking, lwipstack is part of netsockets

0xc0170 committed:
CMake: fix nanostack and netsockets linking libraries

0xc0170 committed:
CMake ppp: move add library to its origin

0xc0170 committed:
CMake mbedtls: fix naming

0xc0170 committed:
CMake nanostack: fix naming libraries, use `-`

0xc0170 committed:
CMake cellular: link each driver separately

I also reverted the previous change to STM component. We do not yet have functionality to
include/exclude drivers in CMake within Mbed OS.

0xc0170 committed:
CMake netsocket: remove unused variable MBED_NETSOCKET_EXTRA_LIBRARIES

0xc0170 committed:
CMake emac: revert back to using labels

0xc0170 committed:
CMake connectivity: revert ble change and cellular interface

0xc0170 committed:
CMake emac stm32: fix missing library

0xc0170 committed:
CMake connectivity: fix linking to mbedtls, header from drivers must be exposed

This is how application gets in mbedtls driver header:

```
In file included from ../../../../mbed-os/connectivity/mbedtls/./include/mbedtls/config.h:31,
                 from ../../../../mbed-os/connectivity/mbedtls/./include/mbedtls/platform.h:37,
                 from ../../../../mbed-os/connectivity/netsocket/./include/netsocket/TLSSocketWrapper.h:28,
                 from ../../../../mbed-os/connectivity/netsocket/./include/netsocket/nsapi.h:45,
                 from ../../../../mbed-os/mbed.h:27,
                 from ../../../../source/main.cpp:17:
../../../../mbed-os/connectivity/mbedtls/./platform/inc/platform_mbed.h:47:10: fatal error: mbedtls_device.h: No such file or directory
   47 | #include "mbedtls_device.h"
```

0xc0170 committed:
CMake netsockets: fix emac drivers to link to emac

hugueskamba commented on connectivity/CMakeLists.txt:
```diff
@@ -7,33 +7,20 @@ add_library(mbed-ble INTERFACE)
 add_library(mbed-ble-blue_nrg INTERFACE)
 add_library(mbed-ble-cordio INTERFACE)
 add_library(mbed-ble-cordio_ll INTERFACE)
-add_library(mbed-cellular INTERFACE)
```
Does it build? I think the cellular drivers would complain with something like:
```
Cannot specify link libraries for target "mbed-cellular" which is not built by this project.
```

If not, it is perhaps because `connectivity/cellular/` is alphabetically before `connectivity/drivers/` thus the `mbed-cellular` CMake target is created by the time CMake reaches `drivers/`.
I noticed that `mbed-ble` in [this PR](https://github.com/ARMmbed/mbed-os/pull/14377/files) remained in `connectivity/CMakeLists.txt` instead of being moved to `connectivity/FEATURE_BLE/CMakeLists.txt`. Is it because `connectivity/FEATURE_BLE/` comes after `connectivity/drivers/` and thus will show the error:
```
Cannot specify link libraries for target "mbed-ble" which is not built by this project.
```

Should we keep all connectivity CMake targets here until maybe we refactor the connectivity directory?

@pan- @evedon

0xc0170 replied:
The thing here is that all drivers for cellular are part of mbed-cellular - how t hey were designed. there is no label ifdef even with the old tools. I'll raise multiple issues about how libraries are being built so teams can look at.

ble drivers are optional therefore I kept it as it was. Same for netsockets.

hugueskamba replied:
What I meant in short is:
`mbed-ble` is defined in `connectivity/CMakeLists.txt` whilst `mbed-cellular` is defined in `connectivity/cellular/CMakeLists.txt`. Why?

0xc0170 replied:
The order matters in here. I'll clean this up with comments about the order there.

0xc0170 replied:
I'll rebase this once BLE gets in (it fixes this  file and the order).

0xc0170 committed:
CMake connectivity: fix nanostack libraries linking

0xc0170 committed:
CMake events: static library

No weak symbols, this can be static library.

This also resolves a problem with object libraries https://gitlab.kitware.com/cmake/cmake/-/issues/18090.

0xc0170 committed:
CMake connectivity: move netsocket/mbedtls add library

0xc0170 committed:
CMake emac: add missing events

0xc0170 committed:
CMake connectivity: fix conflict from rebase

0xc0170 committed:
CMake randlib: STATIC library

To achieve this, we have to take one source from nanostack otherwise
we pull in entire nanostack and hit some deps issue

This will be cleaned up once we look closer at nanostack & netsockets.

Nanostack can be static library, only few files have weak symbols (related to netsockets and Mbed OS).

0xc0170 committed:
CMake mbedtls drivers: link to mbedtls PRIVATE and expose headers

Libraries require mbedtls drivers paths to get mbedtls configuration.
To get driver to mbedtls, we need to link privately, not exposing object files to the consumers of mbed-tls. And provide publicly headers, as drivers are part of mbedtls, they must be also available.

0xc0170 committed:
CMake mbedtls: static library, now weak symbols involved

0xc0170 committed:
CMake netsocket: link publicly mbedtls to expose mbedtls to an application

0xc0170 committed:
CMake netsocket: move tls to be privately linked

0xc0170 committed:
CMake nanostack: cmsis part links to main component, do not need to provide link libs

0xc0170 committed:
CMake netsocket: expose events and tls as they will be needed in the app as well
